File: Prototype/c++/Prototype.cpp
#include <iostream>
#include <string>

class Animal
{
    public:
        virtual void SetName(std::string name){}
        virtual std::string GetName() = 0;
        virtual Animal* clone() const = 0;
};

class Sheep : Animal
{
    private:
        std::string name;
        Sheep(const Sheep& donor) : name(donor.name) {}

    public:
        Sheep() {}
        ~Sheep() {}
        void SetName(std::string name) override;
        std::string GetName() override;
        Sheep* clone() const;
};

int main()
{
    Sheep* sheepDonor = new Sheep();
    std::string name = "sheep";
    sheepDonor->SetName(name);

    Sheep* sheepClone = sheepDonor->clone();

    std::cout << sheepDonor->GetName() << std::endl;
    std::cout << sheepClone->GetName() << std::endl;
    
    return 0;
}

void Sheep::SetName(std::string name)
{
    this->name = name;
}

std::string Sheep::GetName()
{
    return name;
}

Sheep *Sheep::clone() const
{
    return new Sheep(*this);
}

File: Prototype/python/Prototype.py
class Sheep:

    __name: str = ''

    def __init__(self, donor: 'Sheep' = None):
        if donor is not None:
            self.__name = donor.get_name()

    def set_name(self, name: str):
        self.__name = name

    def get_name(self) -> str:
        return self.__name

    def clone(self):
        return Sheep(self)


if __name__ == '__main__':

    sheep_donor: Sheep = Sheep()
    sheep_donor.set_name('sheep')

    sheep_clone: Sheep = sheep_donor.clone()

    print(sheep_donor.get_name())
    print(sheep_clone.get_name())